### PR TITLE
enable to create VPC portfowarding rules with source cidr

### PR DESCRIPTION
**Pull request by RodrigoDLopez**
### Description

Fixes: https://github.com/apache/cloudstack/issues/7483

When a guest network is created, ACS allows the configuration of the firewall and port forwarding, load balancing, and VPN rules for all IPs associated with the network (including the source NAT). However, when creating a VPC, firewall configurations are possible only via ACLs; port forwarding and load balancing rules are only possible for other public IPs assigned to the network. Thus, situations, where it is necessary to combine firewall rules and port forwarding/load balancing in VPCs,  are not possible.

To work around the situation, this PR implements an extension to allow the definition of source CIDR when creating port forwarding rules. If the `cidrlist` parameter is not informed, the current behavior is maintained. This attribute will only be used when creating port forwarding rules in VPCs, since in guest networks it is possible to combine firewall and port forwarding/load balancing rules directly on the public IPs allocated to that network.

```
(scc) > create portforwardingrule protocol=tcp privateport=22 publicport=22 openfirewall=false vmguestip=10.1.1.248 virtualmachineid=334d4279-a0e6-4c0a-a999-37aabda08acd ipaddressid=5cca136d-5425-44fa-b6af-aafc29df269e networkid=b61abc91-ee24-4578-bae6-fa2d06cf9370 cidrlist=192.168.200.170/32
{
  "portforwardingrule": {
    "cidrlist": "192.168.200.170/32",
    "fordisplay": true,
    "id": "210bab98-91f3-4911-8680-5189387a7135",
    "ipaddress": "172.16.200.51",
    "ipaddressid": "5cca136d-5425-44fa-b6af-aafc29df269e",
    "networkid": "b61abc91-ee24-4578-bae6-fa2d06cf9370",
    "privateendport": "22",
    "privateport": "22",
    "protocol": "tcp",
    "publicendport": "22",
    "publicport": "22",
    "state": "Active",
    "tags": [],
    "virtualmachinedisplayname": "pf-01",
    "virtualmachineid": "334d4279-a0e6-4c0a-a999-37aabda08acd",
    "virtualmachinename": "pf-01",
    "vmguestip": "10.1.1.248"
  }
}
```

Furthermore, it was observed that when removing a port forwarding rule, even if the ACS confirmed the removal of said rule, it remained active in the VR. This behavior has changed so that the rule is removed properly.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### How Has This Been Tested?

For testing purposes, I created a VPC: `vpc-01` and a Tier: `vpc-01-tier-01`. Using this tier three instances `PF01`, `PF02`, `PF03` was created and the port forward and ACL rules described below were created.

**Firewall Rules Applied (ACL)**
| Origin           | Start Port    | End Port        | Protocolo      |
|------------------|---------------|-----------------|----------------|
| 0.0.0.0/0        | 22            | 22              | TCP (Allow)    |
| 172.16.200.170/32| 23            | 23              | TCP (Allow)    |
| 172.16.200.170/32| 24            | 24              | TCP (Allow)    |
| 0.0.0.0/0        | -----         | -----           | ALL (Deny)     |

**Port forward rules**
| Private port     | Public port   | Protocol      | Source CIDR   | VM             |
|------------------|---------------|---------------|---            |----------------|
| 22 - 22          | 22 - 22       | TCP           | ---           | PF01           |
| 22 - 22          | 23 - 23       | TCP           | ---           | PF02           |
| 22 - 22          | 24 - 24       | TCP           | ---           | PF03           |

This way, after applying the ACL rule that authorizes access to port `22` from any IP (`0.0.0.0/0`); access to other instances (`PF02` and `PF03`) turns out to be exposed. After applying the port forward rule limiting access to a list of CIDRs, the observed result was that access to the resources/services of instances `PF02` and `PF03` was limited to the CIDR informed when creating the port forward rule

**Regras de port forward**
| Private port     | Public port   | Protocol      | Source CIDR   | VM             |
|------------------|---------------|---------------|---            |----------------|
| 22 - 22          | 22 - 22       | TCP           | ---           | A              |
| 22 - 22          | 23 - 23       | TCP           | 172.16.200.170/32| B              |
| 22 - 22          | 24 - 24       | TCP           | 172.16.200.170/32 | C              |

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@RodrigoDLopez can you please look at the sonar warning? They look serious in this case.

**Comment by weizhouapache** [comment]
Hi @RodrigoDLopez 

Have you tested
- create with a list of cidrs ?
- same operation on isolated networks ?
- create port forwarding rule with same public and same private port, but different cidr
- create port forwarding rule with same public and different private port
- update cidrlist of port forwarding rule ?
- restart vpc with/without cleanup
- restart vpc router

something are missing
- ui changes
- unit tests
- marvin test

Please also ask one of your colleagues to test it.

File: systemvm/debian/opt/cloud/bin/cs/CsNetfilter.py [hunk 1]
```diff
@@ -339,6 +339,9 @@ def to_str(self, delete=False):
                 printable = printable.replace('!_-', '! -')
                 if delete:
                     printable = printable.replace('-A', '-D')
+                    if k == "--save-mark":
```
Comment (weizhouapache) [inline systemvm/debian/opt/cloud/bin/cs/CsNetfilter.py:342]: @RodrigoDLopez 
what's the purpose of this change ?

**Comment by assistanz247** [comment]
@RodrigoDLopez 

Yes, our use case has been covered in the PR. Thanks a lot :-)

**Comment by weizhouapache** [comment]
> @RodrigoDLopez
> 
> Yes, our use case has been covered in the PR. Thanks a lot :-)

cool, good to know it

**Comment by sureshanaparti** [comment]
Hi @RodrigoDLopez  It seems the issue #7483 is fixed with this PR. Can target this PR for 4.19.1?

**Comment by DaanHoogland** [comment]
@RodrigoDLopez can you look at the conflicts?

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java [hunk 2]
```diff
@@ -107,8 +108,13 @@ public class CreatePortForwardingRuleCmd extends BaseAsyncCreateCmd implements P
                 description = "the ID of the virtual machine for the port forwarding rule")
     private Long virtualMachineId;
 
-    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the cidr list to forward traffic from. Multiple entries must be separated by a single comma character (,). This parameter is deprecated. Do not use.")
-    private List<String> cidrlist;
+    @Parameter(name = ApiConstants.CIDR_LIST,
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java:111]: I suggest we use SOURCE_CIDR_LIST

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java [hunk 3]
```diff
@@ -107,8 +108,13 @@ public class CreatePortForwardingRuleCmd extends BaseAsyncCreateCmd implements P
...
+            type = CommandType.LIST,
+            collectionType = CommandType.STRING,
+            description = "the CIDR list to allow traffic, all other CIDRs will be blocked. " +
+                    "Multiple entries must be separated by a single comma character (,). This param will be used only for VPC's networks. By default, all CIDRs are allowed.")
+    private List<String> cidrList;
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java:116]: Also sourceCidrList

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java [hunk 4]
```diff
@@ -107,8 +108,13 @@ public class CreatePortForwardingRuleCmd extends BaseAsyncCreateCmd implements P
                 description = "the ID of the virtual machine for the port forwarding rule")
     private Long virtualMachineId;
 
-    @Parameter(name = ApiConstants.CIDR_LIST, type = CommandType.LIST, collectionType = CommandType.STRING, description = "the cidr list to forward traffic from. Multiple entries must be separated by a single comma character (,). This parameter is deprecated. Do not use.")
-    private List<String> cidrlist;
+    @Parameter(name = ApiConstants.CIDR_LIST,
+            type = CommandType.LIST,
+            collectionType = CommandType.STRING,
+            description = "the CIDR list to allow traffic, all other CIDRs will be blocked. " +
+                    "Multiple entries must be separated by a single comma character (,). This param will be used only for VPC's networks. By default, all CIDRs are allowed.")
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java:115]: I remember there are some changes for source cidr of load balancing rules. Does it work in isolated networks? If yes, this could be extended to support isolated networks as well.

File: systemvm/debian/opt/cloud/bin/cs/CsNetfilter.py (hunk 1 above)
Comment (winterhazel) [reply systemvm/debian/opt/cloud/bin/cs/CsNetfilter.py:394]: Hi Wei,

I talked with Rodrigo, and he said that this change was made to fix an error when deleting port forwarding rules where the VR couldn't find the correct rule to delete. We couldn't figure out the exact situation that resulted in this error though.

From what I've tested so far, this change does not seem necessary (at least anymore). Maybe the problem that caused the error got fixed in another PR. Therefore, I'm going to remove these lines, finish adjusting this PR and do some extensive testing. If it really does not seem to affect anything, I'll leave the PR without these changes.

File: api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java (hunk 2 above)
Comment (winterhazel) [reply api/src/main/java/org/apache/cloudstack/api/command/user/firewall/CreatePortForwardingRuleCmd.java:109]: I'm keeping the param as `cidrlist` for now so that it matches the  `createLoadBalancerRule` API and so that we don't add another parameter or change an existing param's name while we don't have a well-defined protocol to introduce breaking changes. However, I'll address it as "source CIDR list" in the UI and change its description to provide a better idea on what the parameter represents.

**Review by DaanHoogland — APPROVED** [review]
clgtm, needs testing

**Comment by winterhazel** [comment]
@blueorangutan package

**Comment by winterhazel** [comment]
Everything seems to be working as intended now. Below are the tests I did.

## Tests

For each test, I created/updated/deleted port forwarding rules via the UI, and verified that the VR had the expected iptables rules. I also tried to access the public ports via different CIDRs to ensure that the packets were being routed as intended.

### 1. Create PF rule with no source CIDR:

Screenshot from 2024-10-25 17-03-50

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  anywhere             192.168.122.18       tcp dpt:24 to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  anywhere             192.168.122.18       tcp dpt:24 to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 2. Delete PF rule with no source CIDR:

I deleted the previously created rule.

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
```

### 3. Create PF rule with a single source CIDR:

Screenshot from 2024-10-30 11-15-21

```sh
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.200.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.200.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 4. Delete PF rule with a single source CIDR:

I deleted the previously created rule.

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
```

### 5. Create PF rule with a list of source CIDRs:

Screenshot from 2024-10-31 16-03-54

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.122.1        192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.122.1        192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 6. Update source CIDR list of PF rule to another list:

```sh
(admin) 🐢 > update portforwardingrule id=650ec01f-9e5e-4b70-9ff2-ab8c9cff2051  cidrlist="192.168.202.0/24,192.168.203.0/24,192.168.204.10/32"
{
  "portforwardingrule": {
    "cidrlist": "192.168.202.0/24,192.168.203.0/24,192.168.204.10/32",
    "fordisplay": true,
    "id": "650ec01f-9e5e-4b70-9ff2-ab8c9cff2051",
    "ipaddress": "192.168.122.18",
    "ipaddressid": "7118532f-a6be-4943-bc06-037ea00ef4d0",
    "networkid": "49a35e15-ea93-4434-a158-0ffbd3743de1",
    "privateendport": "22",
    "privateport": "22",
    "protocol": "tcp",
    "publicendport": "23",
    "publicport": "23",
    "state": "Active",
    "tags": [],
    "virtualmachinedisplayname": "test-vm",
    "virtualmachineid": "8584e8a2-0759-4aba-885d-8c0b903fdfc2",
    "virtualmachinename": "test-vm",
    "vmguestip": "10.0.0.199"
  }
}
```

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.203.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.204.10       192.168.122.18       tcp dpt:telnet to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.203.0/24     192.168.122.18       tcp dpt:telnet to:10.0.0.199:22
DNAT       tcp  --  192.168.204.10       192.168.122.18       tcp dpt:telnet to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 7. Delete PF rule with a list of source CIDRs:

I deleted the updated rule.

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
```

### 8. Update source CIDR of PF rule to empty

Before the update:

Screenshot from 2024-10-31 18-02-07

```sh
root@r-39-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.200.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:23

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.200.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:23

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:telnet to:10.0.0.1
```

Update:

```sh
(admin) 🐢 > update portforwardingrule id=7340fcb8-4dee-4f74-9bb9-cae92b944783 cidrlist=""
{
  "portforwardingrule": {
    "cidrlist": "",
    "fordisplay": true,
    "id": "7340fcb8-4dee-4f74-9bb9-cae92b944783",
    "ipaddress": "192.168.122.18",
    "ipaddressid": "7118532f-a6be-4943-bc06-037ea00ef4d0",
    "networkid": "49a35e15-ea93-4434-a158-0ffbd3743de1",
    "privateendport": "23",
    "privateport": "23",
    "protocol": "tcp",
    "publicendport": "22",
    "publicport": "22",
    "state": "Active",
    "tags": [],
    "virtualmachinedisplayname": "test-vm",
    "virtualmachineid": "8584e8a2-0759-4aba-885d-8c0b903fdfc2",
    "virtualmachinename": "test-vm",
    "vmguestip": "10.0.0.199"
  }
}
```

```sh
root@r-39-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  anywhere             192.168.122.18       tcp dpt:ssh to:10.0.0.199:23

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  anywhere             192.168.122.18       tcp dpt:ssh to:10.0.0.199:23

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:telnet to:10.0.0.1
```

### 9. Create PF rules with the same public and private ports, but different source CIDRs, for the same VM:

Screenshot from 2024-10-31 16-30-17

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.200.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.122.1        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.200.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.122.1        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 10. Delete one of the previously created rules:

I deleted the rule that had 192.168.200.0/24 as the source CIDR.

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.122.1        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.122.1        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 11. Create PF rules with the same public and private ports, but different source CIDRs, for different VMs:

Screenshot from 2024-10-31 15-34-25

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  10.0.0.0/24          192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.122.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.186:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  10.0.0.0/24          192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.122.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.186:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
SNAT       tcp  --  10.0.0.0/24          test-vm2             tcp dpt:ssh to:10.0.0.1
```

### 12. Create PF rules with the same public port, but different private port for different source CIDRs:

Screenshot from 2024-10-31 15-38-37

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  10.0.0.0/24          192.168.122.18       tcp dpt:ssh to:10.0.0.199:23
DNAT       tcp  --  192.168.122.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  10.0.0.0/24          192.168.122.18       tcp dpt:ssh to:10.0.0.199:23
DNAT       tcp  --  192.168.122.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:telnet to:10.0.0.1
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 13. Delete one of the previously created rules:

I deleted the rule that had `192.168.122.0/24` as source CIDR.

```sh
root@r-36-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  10.0.0.0/24          192.168.122.18       tcp dpt:ssh to:10.0.0.199:23

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  10.0.0.0/24          192.168.122.18       tcp dpt:ssh to:10.0.0.199:23

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:telnet to:10.0.0.1
```

### 14. Restart VPC with clean-up:

I added a rule that had a list of source CIDRs.

Screenshot from 2024-10-31 16-10-44

Then, I restarted the VPC with clean-up. After restarting, I verified the iptables.

```sh
root@r-37-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.203.2        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.203.2        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 15. Restart VPC without clean-up:

I restarted the same VPC, but without clean-up. After restarting, I verified the iptables.

```sh
root@r-37-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.203.2        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.203.2        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 16. Restart VPC router:

I rebooted the VR of the same VPC. After rebooting, I verified the iptables.

```sh
root@r-37-VM:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.203.2        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DNAT       tcp  --  192.168.201.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.202.0/24     192.168.122.18       tcp dpt:ssh to:10.0.0.199:22
DNAT       tcp  --  192.168.203.2        192.168.122.18       tcp dpt:ssh to:10.0.0.199:22

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
SNAT       all  --  anywhere             anywhere             to:192.168.122.17
SNAT       tcp  --  10.0.0.0/24          test-vm              tcp dpt:ssh to:10.0.0.1
```

### 17. Try to create a PF rule with an invalid source CIDR

I verified that I got an exception.

Screenshot from 2024-10-30 20-01-25

### 18. Try update a PF rule's source CIDR to an invalid value

Also got an error.

```sh
(admin) 🐢 > update portforwardingrule id=617b41c6-264c-47bb-bdfa-0bd1fdcf2b01 cidrlist="test"
{
  "account": "admin",
  "accountid": "60b92e66-28b2-11ef-a04e-6ac09c2b3d6b",
  "cmd": "org.apache.cloudstack.api.command.user.firewall.UpdatePortForwardingRuleCmd",
  "completed": "2024-10-31T19:42:51+0000",
  "created": "2024-10-31T19:42:50+0000",
  "domainid": "40815ad7-28b2-11ef-a04e-6ac09c2b3d6b",
  "domainpath": "ROOT",
  "jobid": "08f8132c-c715-4ebf-97ac-5cad81ea0648",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 431,
    "errortext": "The given source CIDR [test] is invalid."
  },
  "jobresultcode": 431,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "60ba931c-28b2-11ef-a04e-6ac09c2b3d6b"
}
```

### 19. Unit tests

I added unit tests for the modified `if` condition, and verified that the condition produced the same results before and after the changes for the cases that did not involve the new parameter.

**Comment by DaanHoogland** [comment]
> Everything seems to be working as intended now. Below are the tests I did.

@weizhouapache , does this cover your concerns?

> Hi @RodrigoDLopez It seems the issue #7483 is fixed with this PR. Can target this PR for 4.19.1?

@winterhazel , will you considder this?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by winterhazel** [comment]
> > Hi @RodrigoDLopez It seems the issue #7483 is fixed with this PR. Can target this PR for 4.19.1?
> 
> @winterhazel , will you considder this?

My bad, I missed this message. Yes, I'll target 4.19.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by winterhazel** [comment]
> @RodrigoDLopez @winterhazel , are you guys testing this?

@DaanHoogland I already tested it in https://github.com/apache/cloudstack/pull/7081#issuecomment-2450828651, and consider this PR as ready for merge.
